program/src/state/rules.rs

```diff
@@ -398,7 +398,15 @@ impl Rule {
                 };
 
                 if let Some(account) = accounts.get(key) {
-                    if *account.owner == *program {
+                    let data = match account.data.try_borrow() {
```

Comment:
Can this borrow bit be nested in the is_zeroed function as well, instead of duplicating it?

Reply:
I tried this but it's not looking as good.  What I wanted to do was:

```
pub fn is_zeroed(account: &AccountInfo) -> Result<bool, ProgramError> {
    ...
    let data = match account.data.try_borrow() {
        ...
```

But then I cannot just call it in the rule using `is_zeroed(&data)?` because `low_level_validate()` returns a tuple not a `Result`.  So I also cannot simply map the return value to another `Result`.

So I end up having to destructure it and translate the error to the tuple manually, so it ends up looking something like:
```
let is_zeroed = match is_zeroed(account) {
    Ok(val) => val,
    Err(err) => return (false, err),
};

if is_zeroed {
    // Account must have nonzero data to count as program-owned.
    return (false, self.to_error());
} else if *account.owner == *program {
    return (true, self.to_error());
}
```

Which I actually find more confusing to look at than what was there originally.

Reply:
Yeah, I agree. Just curious! I think it's good the way it was, then.